pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+// The function performs one pass over both slices, sequentially.
+func destinationsMatch(destinations []string, placementDecisionDestinations []v1alpha1.Destination) bool {
+	if len(destinations) != len(placementDecisionDestinations) {
+		return false
+	}
+
+	// calculate match by counting matching objects then comparing to length.
+	// this works since entries are unique.
+	matches := map[string]interface{}{}
+	matchesCounter := 0
+
+	// fill up matches
+	for _, destination := range placementDecisionDestinations {
+		matches[destination.ClusterId] = nil
+	}
+
+	// check matches
+	for _, destination := range destinations {
+		if _, found := matches[destination]; !found {
+			return false
+		}
+
+		matchesCounter++
+	}
+
+	return len(placementDecisionDestinations) == matchesCounter
+}
+
+// workloadMatchesPlacementDecisionSpec returns true if the workload in the
+// resolution matches the workload in the placement decision spec.
+func workloadMatchesPlacementDecisionSpec(placementDecisionSpecWorkload *v1alpha1.DownsyncObjectReferences,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	// check lengths
+	clusterScopedObjectsCount := 0
+	for _, clusterScopeDownsyncObjects := range placementDecisionSpecWorkload.ClusterScope {
+		clusterScopedObjectsCount += len(clusterScopeDownsyncObjects.ObjectNames)
+	}
+
+	namespacedObjectsCount := 0
+	for _, namespaceScopeDownsyncObjects := range placementDecisionSpecWorkload.NamespaceScope {
+		for _, objectsByNamespace := range namespaceScopeDownsyncObjects.ObjectsByNamespace {
+			namespacedObjectsCount += len(objectsByNamespace.Names)
+		}
+	}
+
+	if len(objectIdentifierToKeyMap) != clusterScopedObjectsCount+namespacedObjectsCount {
+		return false
+	}
+
+	// again we can check match by making sure all objects are mapped, since entries are unique and the length is equal.
+	// check cluster-scoped all exist
+	if !placementDecisionClusterScopeIsMapped(placementDecisionSpecWorkload.ClusterScope, objectIdentifierToKeyMap,
+		gvkGvrMapper) {
+		return false
+	}
+
+	// check namespace-scoped all exist
+	return namespaceScopeMatchesPlacementDecisionSpec(placementDecisionSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
+		gvkGvrMapper)
+}
+
+// clusterScopeMatchesPlacementDecisionSpec returns true if the cluster-scope
+// section in the placement decision spec all exist in the resolution.
+func placementDecisionClusterScopeIsMapped(placementDecisionSpecClusterScope []v1alpha1.ClusterScopeDownsyncObjects,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	for _, clusterScopeDownsyncObjects := range placementDecisionSpecClusterScope {
+		for _, objName := range clusterScopeDownsyncObjects.ObjectNames {
+			gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
+			gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+			if !found {
+				return false // assume obj not mapped
```

Comment:
The reader wonders why it is valid to simply make an assumption here. If I understand correctly, what is really going on here is that it is always safe to return `false`. The reader needs an explanation, probably centered at the interface definition.